feature/pokedex -> dev

alexanderboswell opened the pull request:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/byu-oit/ios-byusuite/234)
<!-- Reviewable:end -->

romrell4 commented:
Reviewed 1 of 8 files at r1.
Review status: 1 of 8 files reviewed at latest revision, 2 unresolved discussions.

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 16 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsQDwD-HTITm4kl8qNa:-KsQDwD-HTITm4kl8qNb:bq258l0) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L16)):*
> ```Swift
>     
>     @IBOutlet weak var imageView: UIImageView!
>     
> ```

You can remove the newlines

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 11 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsQEAIaTwm7d-Sf0pnt:-KsQEAIaTwm7d-Sf0pnu:b72tsrb) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L11)):*
> ```Swift
> import UIKit
> 
> class PokemonListViewController: ByuViewController2, UITableViewDataSource {
> ```

I think you should be able to use the TableDataViewController

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 7 of 8 files at r1.
Review status: all files reviewed at latest revision, 38 unresolved discussions.

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 15 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdMw_FkqWbN97rQnJZ:-KsdMw_FkqWbN97rQnJ_:b-qrumgs) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L15)):*
> ```Swift
>     //MARK: Outlets
>     
>     @IBOutlet weak var imageView: UIImageView!
> ```

These can all be made private

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 29 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdMsVxSgIsOc3ESaAp:-KsdMsVxSgIsOc3ESaAq:bhniu6i) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L29)):*
> ```Swift
>     @IBOutlet weak var spinner: UIActivityIndicatorView!
>     
>     //MARK: Properties
> ```

These should be marked as `Public Properties`

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 46 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdNCzrDQTjSzDK4GJf:-KsdNCzsRscrymSLMw1i:b-8ic57t) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L46)):*
> ```Swift
>     func loadPokemon() {
>         if let pokemonName = pokemonRef?.name {
>         PokemonClient.getPokemon(pokemonName: pokemonName, completionHandler: { (pokemon, error) in
> ```

I think the indenting in this file is off... Select the whole file and do a `CTRL-i`

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 51 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdN1zHkRbA0kib0aCs:-KsdN1zHkRbA0kib0aCt:b-suilqs) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L51)):*
> ```Swift
>                 self.pokemon = pokemon
>                 if let weight = pokemon.weight, let height = pokemon.height, let type = pokemon.type, let id = pokemon.id, let imageUrl = pokemon.imageUrl {
>                     self.weightLabel.text = "Weight: " + String(weight)
> ```

Use swift string formatting:
`self.weightLabel.text = "Weight: \(weight)"`

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 55 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdNPMtres86iPs7tmA:-KsdNPMtres86iPs7tmB:blyxplz) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L55)):*
> ```Swift
>                     self.typeLabel.text = "Types: " + type
>                     self.idLabel.text = "ID: " + String(id)
>                     PokemonClient.loadImage(imageUrl: imageUrl, completionHandler: { (image, error) in
> ```

No need for a client function

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 58 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdNRtn9PZIeShKG9Gh:-KsdNRtn9PZIeShKG9Gi:b-g93ccy) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L58)):*
> ```Swift
>                         if let image = image {
>                             self.imageView.image = image
>                             PokemonClient.getFlavorText(pokemonName: pokemonName, completionHandler: { (flavorText, error) in
> ```

I don't think you need to wait for the image to come back before loading the flavor text. That would make the load time longer

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 28 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdC6aR1tJHb4EgHiLT:-KsdC6aR1tJHb4EgHiLU:b752ciu) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L28)):*
> ```Swift
>         super.viewDidLoad()
> 
>         self.tableView.tableFooterView?.isHidden = true
> ```

No need for self here. You'll notice in our code base that we have been rather inconsistent with this (because obj-c requires the `self.` before properties), but we've been trying to follow this standard: `Only use self when the reference is needed to understand what resource is being accessed (e.g. in a closure). Other uses of self may be necessary for ambiguous properties or for clarity, but most references of self may be inferred rather than explicit`

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 28 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdCbbGV-L4xUZcJmu1:-KsdCbbGV-L4xUZcJmu2:bsjdfao) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L28)):*
> ```Swift
>         super.viewDidLoad()
> 
>         self.tableView.tableFooterView?.isHidden = true
> ```

I believe if you use the ByuTableDataViewController, it will handle the table footer for you

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 55 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdCtE0g-FOV-beiZHU:-KsdCtE0g-FOV-beiZHV:b-gdh5h1) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L55)):*
> ```Swift
>     }
>     
>     //MARK: UITableViewDataSource/Delegate Methods
> ```

We normally order these in a top-down structure. Here are some examples:
```swift
numberOfSections
titleForHeaderInSection
viewForHeaderInSection
titleForFooterInSection
viewForFooterInSection

numberOfRowsInSection
heightForRowAtIndexPath
cellForRowAtIndexPath
didSelectRowAtIndexPath
didDeselectRowAtIndexPath

etc
```
Try reordering your methods to make them more top-down

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 61 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdCox1iZgL_iV5wPg2:-KsdCox1iZgL_iV5wPg3:blrrapr) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L61)):*
> ```Swift
>     }
>     func numberOfSections(in tableView: UITableView) -> Int {
>         return 2
> ```

Comment here about why there are two sections

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 84 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdE4buv2IcWMHHOsrX:-KsdE4buv2IcWMHHOsrY:b-et69i9) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L84)):*
> ```Swift
>     
>     func tableView(_ tableView: UITableView, willDisplayFooterView view: UIView, forSection section: Int) {
>         if let nextApiCall = nextApiCall {
> ```

Combine these two ifs

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 91 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdEE3BkbLovsKEWt4j:-KsdEE3BkbLovsKEWt4k:bc1765w) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L91)):*
> ```Swift
>                         
>                         for pokemonRef in pokemonRefs {
>                             self.pokemonRefs.append(pokemonRef)
> ```

I believe there is an `append(contentsOf:)` that will do this more efficiently

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 107 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdET1XplA8EoZr2pTv:-KsdET1XplA8EoZr2pTw:b509tzc) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L107)):*
> ```Swift
>     
>     func loadPokemonRefs() {
>         
> ```

Remove excess line

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 108 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdEYLtbVXd6nL3klTa:-KsdEYLtbVXd6nL3klTb:be28h5j) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L108)):*
> ```Swift
>     func loadPokemonRefs() {
>         
>         PokemonClient.getPokemonRefs { (pokemonRefs, nextApiCall, error) in
> ```

There's a lot of duplicated logic here... Could the other one call this function instead of duplicating?

---

*[byuSuite/Apps/Pokedex/model/Pokemon.swift, line 11 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdEfMhzv2bX0RJmJxj:-KsdEfMhzv2bX0RJmJxk:b-rgqdi8) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/Pokemon.swift#L11)):*
> ```Swift
> import UIKit
> 
> class Pokemon: NSObject {
> ```

What is your opinion about Swift classes and extending NSObject versus not extending at all? I haven't found much documentation that leans either way

---

*[byuSuite/Apps/Pokedex/model/Pokemon.swift, line 14 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdEovam9BMSf1v1PLH:-KsdEovam9BMSf1v1PLI:bkwrgfz) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/Pokemon.swift#L14)):*
> ```Swift
>     
>     var name: String?
>     
> ```

No need for extra lines. Remove these in other model classes as well

---

*[byuSuite/Apps/Pokedex/model/Pokemon.swift, line 31 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdJCVUdLHZTie7MMLZ:-KsdJCVUdLHZTie7MML_:bd7joia) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/Pokemon.swift#L31)):*
> ```Swift
>         weight = dictionary["weight"] as? Int
>         height = dictionary["height"] as? Int
>         var tempType = ""
> ```

I think there should be a simpler way of doing this. Talk to me about this one

---

*[byuSuite/Apps/Pokedex/model/Pokemon.swift, line 47 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdIh3-tu0lyLJgBUtw:-KsdIh3-tu0lyLJgBUtx:ba0wxb6) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/Pokemon.swift#L47)):*
> ```Swift
>         type = tempType
>         id = dictionary["id"] as? Int
>         imageUrl = ""
> ```

Why are we setting these to empty strings? If we actually want to do this, then let's default these at the top (talk to me for explanation)

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 12 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdJJhvF3vvhsiaiiCC:-KsdJJhvF3vvhsiaiiCD:b4r1ur7) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L12)):*
> ```Swift
> 
> class PokemonClient: ByuClient2 {
>     private static let POKEMON_BASE_URL = "https://pokeapi.co/api/v2/pokemon/"
> ```

These should be defined outside of the class (then you don't have to mark them as static)

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 12 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdJNERms5s5dslg-UT:-KsdJNERms5s5dslg-UU:bkog2fn) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L12)):*
> ```Swift
> 
> class PokemonClient: ByuClient2 {
>     private static let POKEMON_BASE_URL = "https://pokeapi.co/api/v2/pokemon/"
> ```

There should be one base_url, which would be "https://pokeapi.co/api/v2". The other portions are part of the pathParams. For instance, to get this first url, we should use `super.url(base: url, pathParams: ["pokemon"])`

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 16 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdKilPbCt9dF0WGGHl:-KsdKilPbCt9dF0WGGHm:b-k8tmce) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L16)):*
> ```Swift
>     
>     // optional url parameter so that you can pass in the offset api call else by default it do a call to the base url
>     static func getPokemonRefs(url: String? = POKEMON_BASE_URL, completionHandler: @escaping ([PokemonRef]?, String, ByuError?) -> Void) {
> ```

Rename the completionHandler to `callback`

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 16 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdKqHtKZj059oZjgy7:-KsdKqHtKZj059oZjgy8:bjmlpse) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L16)):*
> ```Swift
>     
>     // optional url parameter so that you can pass in the offset api call else by default it do a call to the base url
>     static func getPokemonRefs(url: String? = POKEMON_BASE_URL, completionHandler: @escaping ([PokemonRef]?, String, ByuError?) -> Void) {
> ```

The second parameter of the callback should be a `String?` in case there isn't a nextUrl

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 17 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdJq02zFMg475ERh4K:-KsdJq0300YaOg2bEpg9:b-cwrgw) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L17)):*
> ```Swift
>     // optional url parameter so that you can pass in the offset api call else by default it do a call to the base url
>     static func getPokemonRefs(url: String? = POKEMON_BASE_URL, completionHandler: @escaping ([PokemonRef]?, String, ByuError?) -> Void) {
>         if let url = url {
> ```

It looks like we never handle the scenario where they pass in a null url...

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 20 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdKCw5rUZmZdxbgGL1:-KsdKCw5rUZmZdxbgGL2:b-fi9yyi) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L20)):*
> ```Swift
>             let request = ByuRequest2(url: super.url(base: url), authMethod: .NO_AUTH)
>             request.startAsynchronous{ (response) in
>                 if response.failed {
> ```

We recently switched to always check the success scenario first. So these callbacks would look like this:
```swift
if response.succeeded, let responseDict = response.getDataJson() as? [String: Any], let results = responseDict["results"] as? [[String: Any]] {
  ...
} else {
  callback(nil, nil, response.error)
}
```


---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 40 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdK-z6rK_ty7AU3ZH9:-KsdK-z6rK_ty7AU3ZHA:b-glmhu9) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L40)):*
> ```Swift
>     }
>     
>     static func getPokemon(pokemonName: String, completionHandler: @escaping (Pokemon?, ByuError?) -> Void) {
> ```

I don't remember a lookup by name... We should probably use the id here instead (more reliable)

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 40 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdL-x6hg8aEjoo1Gs8:-KsdL-x6hg8aEjoo1Gs9:b-xxolh) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L40)):*
> ```Swift
>     }
>     
>     static func getPokemon(pokemonName: String, completionHandler: @escaping (Pokemon?, ByuError?) -> Void) {
> ```

Rename completionHandler to callback

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 41 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdK4QGPdUFa0T2xqM3:-KsdK4QGPdUFa0T2xqM4:b4hzp2s) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L41)):*
> ```Swift
>     
>     static func getPokemon(pokemonName: String, completionHandler: @escaping (Pokemon?, ByuError?) -> Void) {
>         let request = ByuRequest2(url: super.url(base: POKEMON_BASE_URL + pokemonName), authMethod: .NO_AUTH)
> ```

This should be `super.url(base: BASE_URL, pathParams:["pokemon", pokemonId])`

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 43 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdL5k64yoICBGTw2Sv:-KsdL5k64yoICBGTw2Sw:byhzwyf) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L43)):*
> ```Swift
>         let request = ByuRequest2(url: super.url(base: POKEMON_BASE_URL + pokemonName), authMethod: .NO_AUTH)
>         request.startAsynchronous{ (response) in
>             if response.failed {
> ```

Flop if/else

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 56 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLQY97wdw82vlTnwU:-KsdLQY97wdw82vlTnwV:bmmohox) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L56)):*
> ```Swift
>     }
>     
>     static func getFlavorText(pokemonName: String, completionHandler: @escaping(String?, ByuError?) -> Void) {
> ```

Rename completionHandler

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 57 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLBKCropNAobsNkHn:-KsdLBKCropNAobsNkHo:b-3zzbsn) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L57)):*
> ```Swift
>     
>     static func getFlavorText(pokemonName: String, completionHandler: @escaping(String?, ByuError?) -> Void) {
>         let request = ByuRequest2(url: super.url(base: POKEMON_SPECIES_BASE_URL + pokemonName), authMethod: .NO_AUTH)
> ```

use pathParams

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 59 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLDAw-pnngvMb7Agg:-KsdLDAw-pnngvMb7Agh:bvs5ylz) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L59)):*
> ```Swift
>         let request = ByuRequest2(url: super.url(base: POKEMON_SPECIES_BASE_URL + pokemonName), authMethod: .NO_AUTH)
>         request.startAsynchronous { (response) in
>             if response.failed {
> ```

flop if/else

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 67 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLIAKX0khbARdU7-O:-KsdLIAKX0khbARdU7-P:b7t8d28) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L67)):*
> ```Swift
>                 }
>                 if let flavorTextEntries = responseDict["flavor_text_entries"] as? [[String: Any]] {
>                     for i in flavorTextEntries {
> ```

We can simplify this with a `filter` of `first(where:)`

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 85 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLSaGBgfyGyNzur1T:-KsdLSaGBgfyGyNzur1U:b-4i0hj5) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L85)):*
> ```Swift
>     }
>     
>     static func loadImage( imageUrl: String, completionHandler: @escaping (UIImage?, ByuError?) -> Void) {
> ```

Remove extra space

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 85 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLUIERFr9cqOdJYmQ:-KsdLUIERFr9cqOdJYmR:bmmohox) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L85)):*
> ```Swift
>     }
>     
>     static func loadImage( imageUrl: String, completionHandler: @escaping (UIImage?, ByuError?) -> Void) {
> ```

Rename completionHandler

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 85 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLap_0IW0v45yOqHv:-KsdLap_0IW0v45yOqHw:b-3zj1my) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L85)):*
> ```Swift
>     }
>     
>     static func loadImage( imageUrl: String, completionHandler: @escaping (UIImage?, ByuError?) -> Void) {
> ```

I don't think you'll need a Client method for this. Look into `UIImage(data: Data(withContentsOfUrl:))`

---

*[byuSuite/Apps/Pokedex/view/Pokedex.storyboard, line 1 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdMR6ZS_PcBg2hGzwS:-KsdMR6ZS_PcBg2hGzwT:b284cxb) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/view/Pokedex.storyboard#L1)):*
> ```XML
> <?xml version="1.0" encoding="UTF-8"?>
> ```

Reorder items in the storyboard for easier navigation

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234)*
<!-- Sent from Reviewable.io -->

alexanderboswell commented:
Updated per review, added id to tableView detailText, capitalized in correct places, and put labels on weight and height. 

---

Review status: 2 of 8 files reviewed at latest revision, 38 unresolved discussions.

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 15 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdMw_FkqWbN97rQnJZ:-KsdQIaWNcIEu3nElOw2:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L15)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

These can all be made private
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 16 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsQDwD-HTITm4kl8qNa:-KsdQJjl5_m1YKgzMHAk:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L16)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

You can remove the newlines
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 29 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdMsVxSgIsOc3ESaAp:-KsdQef6s6_WYs3vdDJn:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L29)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

These should be marked as `Public Properties`
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 46 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdNCzrDQTjSzDK4GJf:-KsdQfF9UCixNm6dnjOw:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L46)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I think the indenting in this file is off... Select the whole file and do a `CTRL-i`
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 51 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdN1zHkRbA0kib0aCs:-KsdQraCA5OtMg2lIg-X:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L51)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Use swift string formatting:
`self.weightLabel.text = "Weight: \(weight)"`
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 55 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdNPMtres86iPs7tmA:-KsdR8tSYpmnZ4seThbu:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L55)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

No need for a client function
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 58 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdNRtn9PZIeShKG9Gh:-KsdRIR6QrXoRAKyUQjV:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L58)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I don't think you need to wait for the image to come back before loading the flavor text. That would make the load time longer
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 11 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsQEAIaTwm7d-Sf0pnt:-KsdfAM1Z9kMrQp4QF1u:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L11)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I think you should be able to use the TableDataViewController
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 28 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdC6aR1tJHb4EgHiLT:-KsdRr9s0Sq5PTYcGl1m:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L28)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

No need for self here. You'll notice in our code base that we have been rather inconsistent with this (because obj-c requires the `self.` before properties), but we've been trying to follow this standard: `Only use self when the reference is needed to understand what resource is being accessed (e.g. in a closure). Other uses of self may be necessary for ambiguous properties or for clarity, but most references of self may be inferred rather than explicit`
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 28 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdCbbGV-L4xUZcJmu1:-Ksdf7Soi4zbCxJwY1mq:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L28)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I believe if you use the ByuTableDataViewController, it will handle the table footer for you
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 55 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdCtE0g-FOV-beiZHU:-KsdU4M04GgvFKgwfU-8:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L55)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

We normally order these in a top-down structure. Here are some examples:
```swift
numberOfSections
titleForHeaderInSection
viewForHeaderInSection
titleForFooterInSection
viewForFooterInSection

numberOfRowsInSection
heightForRowAtIndexPath
cellForRowAtIndexPath
didSelectRowAtIndexPath
didDeselectRowAtIndexPath

etc
```
Try reordering your methods to make them more top-down
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 61 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdCox1iZgL_iV5wPg2:-KsdfydROh9ApjL0DJ1p:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L61)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Comment here about why there are two sections
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 84 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdE4buv2IcWMHHOsrX:-KsdfxqmA56n0tayd7Nc:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L84)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Combine these two ifs
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 91 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdEE3BkbLovsKEWt4j:-Ksdg19hzTtOG24DrqH1:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L91)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I believe there is an `append(contentsOf:)` that will do this more efficiently
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 107 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdET1XplA8EoZr2pTv:-KsdgAq0Tx7mbj6IzLal:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L107)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Remove excess line
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift, line 108 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdEYLtbVXd6nL3klTa:-KsdgC7f6zgaHLWBWXha:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/controller/PokemonListViewController.swift#L108)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

There's a lot of duplicated logic here... Could the other one call this function instead of duplicating?
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/Pokemon.swift, line 11 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdEfMhzv2bX0RJmJxj:-KseFHjNEqJHytPu8yFF:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/Pokemon.swift#L11)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

What is your opinion about Swift classes and extending NSObject versus not extending at all? I haven't found much documentation that leans either way
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/Pokemon.swift, line 14 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdEovam9BMSf1v1PLH:-KsdSQbjvQg4xtwjmSnY:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/Pokemon.swift#L14)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

No need for extra lines. Remove these in other model classes as well
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/Pokemon.swift, line 31 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdJCVUdLHZTie7MMLZ:-Kse3uf0VmUon0SljkbV:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/Pokemon.swift#L31)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I think there should be a simpler way of doing this. Talk to me about this one
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/Pokemon.swift, line 47 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdIh3-tu0lyLJgBUtw:-KsdvrACvAhJCL8oN6q_:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/Pokemon.swift#L47)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Why are we setting these to empty strings? If we actually want to do this, then let's default these at the top (talk to me for explanation)
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 12 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdJJhvF3vvhsiaiiCC:-Kse4Ae-o-QFN1H1Qq2J:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L12)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

These should be defined outside of the class (then you don't have to mark them as static)
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 12 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdJNERms5s5dslg-UT:-KsdnqdT53_03lmBbJCd:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L12)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

There should be one base_url, which would be "https://pokeapi.co/api/v2". The other portions are part of the pathParams. For instance, to get this first url, we should use `super.url(base: url, pathParams: ["pokemon"])`
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 16 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdKilPbCt9dF0WGGHl:-KsdjBWMglgpx4EI5pbT:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L16)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Rename the completionHandler to `callback`
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 16 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdKqHtKZj059oZjgy7:-Ksdh8RChu18KQVMS8hs:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L16)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

The second parameter of the callback should be a `String?` in case there isn't a nextUrl
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 17 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdJq02zFMg475ERh4K:-KsdhEAdzqJl1mFwYDol:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L17)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

It looks like we never handle the scenario where they pass in a null url...
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 20 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdKCw5rUZmZdxbgGL1:-KsdjmTZdoWQS4h63ifw:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L20)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

We recently switched to always check the success scenario first. So these callbacks would look like this:
```swift
if response.succeeded, let responseDict = response.getDataJson() as? [String: Any], let results = responseDict["results"] as? [[String: Any]] {
  ...
} else {
  callback(nil, nil, response.error)
}
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 40 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdK-z6rK_ty7AU3ZH9:-KsdrHAqDqCAMNvsYbs4:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L40)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I don't remember a lookup by name... We should probably use the id here instead (more reliable)
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 40 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdL-x6hg8aEjoo1Gs8:-KsdhWcll3zdRQWJjTS0:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L40)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Rename completionHandler to callback
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 41 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdK4QGPdUFa0T2xqM3:-KsdrGKEy7f1BwYtSKiX:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L41)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

This should be `super.url(base: BASE_URL, pathParams:["pokemon", pokemonId])`
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 43 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdL5k64yoICBGTw2Sv:-KsdkCnzMLDISy5vvIAG:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L43)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Flop if/else
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 56 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLQY97wdw82vlTnwU:-KsdhXe0_RWQr3xbP6Ai:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L56)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Rename completionHandler
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 57 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLBKCropNAobsNkHn:-KsdnyoTEyeeYISgIQs9:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L57)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

use pathParams
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 59 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLDAw-pnngvMb7Agg:-KsdkDy-ctClSax7F65v:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L59)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

flop if/else
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 67 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLIAKX0khbARdU7-O:-Kse3nkooAexLtkqn1Bl:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L67)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

We can simplify this with a `filter` of `first(where:)`
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 85 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLSaGBgfyGyNzur1T:-Ksdi0rKMvPmdwG5dwTd:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L85)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Remove extra space
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 85 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLUIERFr9cqOdJYmQ:-KsdhzDJKGD3kVBEh3H3:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L85)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Rename completionHandler
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 85 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdLap_0IW0v45yOqHv:-KseFNe60AJcNSIxs5Tm:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L85)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I don't think you'll need a Client method for this. Look into `UIImage(data: Data(withContentsOfUrl:))`
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/view/Pokedex.storyboard, line 1 at r1](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KsdMR6ZS_PcBg2hGzwS:-Kse3pgsBsNN5WbcDRhF:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/89e75705f6f18a2a003cd091e1d75e3787299261/byuSuite/Apps/Pokedex/view/Pokedex.storyboard#L1)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Reorder items in the storyboard for easier navigation
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-:-KseFThaQ3uK6mhfCToL:b-wx90ld)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 1 of 6 files at r2.
Review status: 3 of 8 files reviewed at latest revision, 32 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 5 of 6 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions.

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 66 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseRJCbKDaa7X2Tp36b:-KseRJCbKDaa7X2Tp36c:b-hbvzq8) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L66)):*
> ```Swift
>     }
>     
>     func loadImage( imageUrl: String, callback: @escaping (UIImage?, ByuError?) -> Void) {
> ```

Do you still need this?

---

*[byuSuite/Apps/Pokedex/model/Pokemon.swift, line 31 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseRsww0Fd45G93k4HQ:-KseRsww0Fd45G93k4HR:bdnqaet) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/Pokemon.swift#L31)):*
> ```Swift
>         type = types.joined(separator: ", ")
>         
>         if let id = dictionary["id"] as? Int {
> ```

I think we can simplify this... Does this not work? 
```swift
id = dictionary["id"] as? String
```

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 11 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseS1DO7VMIcszy8E2H:-KseS1DO7VMIcszy8E2I:b26e376) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L11)):*
> ```Swift
> import UIKit
> 
> private  let BASE_URL = "https://pokeapi.co/api/v2"
> ```

Remove excessive space

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 46 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseSdlhFmHjmeWk_FyF:-KseSdlhFmHjmeWk_FyG:b-acpdyv) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L46)):*
> ```Swift
>     
>     static func getFlavorText(pokemonId: String, callback: @escaping(String?, ByuError?) -> Void) {
>         let request = ByuRequest2(url: super.url(base: BASE_URL, pathParams: ["pokemon-species", pokemonId]), authMethod: .NO_AUTH)
> ```

We have some duplicate code creating these requests... Let's write a `makeRequest` function to get rid of the duplicate code

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 50 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseSVqcjv-7PZyF7OLs:-KseSVqcjv-7PZyF7OLt:b-gzwh9t) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L50)):*
> ```Swift
>             if response.succeeded, let responseDict = response.getDataJson() as? [String: Any],
>                 let array = responseDict["flavor_text_entries"] as? [[String: Any]] {
>                 if let flavorTextEntry: [String: Any] = array.first(where: { ($0["language"] as? [String: Any])?["name"] as? String == "en" } ) {
> ```

I think there's a way we can do this with `keypath:` Talk to me on this one

---

*[byuSuite/Apps/Pokedex/model/PokemonRef.swift, line 16 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseSogi9Y9aMHQytUTs:-KseSogi9Y9aMHQytUTt:b-ugjszs) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonRef.swift#L16)):*
> ```Swift
>     var url: String?
>     var id: String {
>         if let url = url , let id = url.characters.split(whereSeparator: { $0 == "/" }).map(String.init).last {
> ```

Remove extra space before ","

---

*[byuSuite/Apps/Pokedex/model/PokemonRef.swift, line 16 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseT-jU5avg0JP1SPSW:-KseT-jU5avg0JP1SPSX:b-a5f3c5) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonRef.swift#L16)):*
> ```Swift
>     var url: String?
>     var id: String {
>         if let url = url , let id = url.characters.split(whereSeparator: { $0 == "/" }).map(String.init).last {
> ```

There's a split function that will do this for you, rather than sending it to characters

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234)*
<!-- Sent from Reviewable.io -->

alexanderboswell commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift, line 66 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseRJCbKDaa7X2Tp36b:-KseTycKdoSzr1c8zSOI:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/controller/PokemonDetailViewController.swift#L66)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Do you still need this?
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/Pokemon.swift, line 31 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseRsww0Fd45G93k4HQ:-KseWutpYbjMVfguqmp1:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/Pokemon.swift#L31)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I think we can simplify this... Does this not work? 
```swift
id = dictionary["id"] as? String
```
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 11 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseS1DO7VMIcszy8E2H:-KseVfLF-Z5Wu7kq-qvp:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L11)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Remove excessive space
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 46 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseSdlhFmHjmeWk_FyF:-Kseds1f3uW0r58olHgw:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L46)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

We have some duplicate code creating these requests... Let's write a `makeRequest` function to get rid of the duplicate code
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 50 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseSVqcjv-7PZyF7OLs:-Ksedspn4_biWnBwvQLU:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L50)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

I think there's a way we can do this with `keypath:` Talk to me on this one
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonRef.swift, line 16 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseSogi9Y9aMHQytUTs:-KseVaf6S9vKrz5U5no6:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonRef.swift#L16)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Remove extra space before ","
</blockquote></details>

Done.

---

*[byuSuite/Apps/Pokedex/model/PokemonRef.swift, line 16 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseT-jU5avg0JP1SPSW:-KseVbF9AXRVAzhrf7RI:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonRef.swift#L16)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

There's a split function that will do this for you, rather than sending it to characters
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 4 of 4 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 11 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseS1DO7VMIcszy8E2H:-KseeRzaFBuCgIEOsJbf:b5f22am) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L11)):*
<details><summary><i>Previously, alexanderboswell (bozzy) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

Not done? Space right before the let

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234)*
<!-- Sent from Reviewable.io -->

alexanderboswell commented:
Review status: 7 of 8 files reviewed at latest revision, 1 unresolved discussion.

---

*[byuSuite/Apps/Pokedex/model/PokemonClient.swift, line 11 at r2](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234#-KseS1DO7VMIcszy8E2H:-KsefkA49abgbV3qYCIx:b-896fix) ([raw file](https://github.com/byu-oit/ios-byusuite/blob/c84f04bb1426e1260e26c01877014e29da7af5dd/byuSuite/Apps/Pokedex/model/PokemonClient.swift#L11)):*
<details><summary><i>Previously, romrell4 (Eric Romrell) wrote…</i></summary><blockquote>

Not done? Space right before the let
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234)*
<!-- Sent from Reviewable.io -->

romrell4 commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/byu-oit/ios-byusuite/234)*
<!-- Sent from Reviewable.io -->
